### PR TITLE
feat(web): gesture path snapshots 🐵

### PR DESCRIPTION
**Pull request by jahorton**
Continuing from point 3 of #9338, this PR allows the matching algorithm to (effectively) segment the touch-contact paths corresponding to gestures as the paths and/or gestures change state.

1.  During gesture-matching and gesture-selection work, the need to allow "resetting" a path - to observe only the new portions - became quite evident.  Some potential gesture specifications may want this reset while others do not... so the best way forward appears to be providing "subviews" into the path.  

2. These "subviews" will be updated when their base "source" receives updates until "disconnected" at a later time.  This is done via event-handlers attached to the original, base "source" that the "subview" aliases.

3.  When a ContactModel is fully matched, a snapshot of the path segment corresponding to the match is made and the "subview" is "disconnected".  This facilitates conversion of the corresponding "matcher" into something of a "match history frame" / snapshot that may be referred to by later stages of the gesture / ongoing touchpath.

@keymanapp-test-bot skip

### CODE REVIEW

**Review by mcdurdin — APPROVED** [review]
LGTM

File: common/web/gesture-recognizer/src/engine/headless/simpleGestureSource.ts [hunk 1]
```diff
@@ -117,4 +133,84 @@ export class SimpleGestureSource<HoveredItemType> {
...
     return jsonClone;
   }
+}
+
+export class SimpleGestureSourceSubview<HoveredItemType> extends SimpleGestureSource<HoveredItemType> {
+  private _baseSource: SimpleGestureSource<HoveredItemType>
+  private subviewDisconnector: () => void;
+
+  /**
+   * Constructs a new SimpleGestureSource instance for tracking updates to an active input point over time.
```
Comment (jahorton) [inline common/web/gesture-recognizer/src/engine/headless/simpleGestureSource.ts:143]:
```suggestion
   * Constructs a new "Subview" into an existing SimpleGestureSource instance.  Future updates of the base
   *  SimpleGestureSource will automatically be included until this instance's `disconnect` method is called.
```

**Comment by jahorton** [comment]
Regarding the review re-request:

Work on the layer to come after #9320 has led me to take another look at this and redo it with a more generalizable solution.  Rather than assume that _all_ gesture-matchers want to look at the same subset of the path as every one of the other matchers... the new approach will allow each one to maintain a "subview" that is independent of those held by the others.  

Unfortunately, it turns out the first attempt's workaround led to a snowball effect once I started working on gesture-model-selection.  At that layer, we care less about the specific types of gesture supported than the abstract properties of potential gesture models.  When viewed in that manner... some gestures can "lock in" and chain only to a single followup, while others can _optionally_ chain to a specific followup or instead terminate if a different gesture matches either the next part of the gesture path or the 'total' path until that point.  The combination of optional lock-in (multitap) and either partial or full path relevance (full path = flick) leads to heavy complications / caveats in the model-selection abstraction without the new subview approach.

-----

Example case: if a longpress-matcher wants to reset its path (due to new base key) before a flick-matcher's distance threshold has been passed, we don't want to reset the flick's distance "progress" - let alone the matcher - but we _do_ want to reset both for the longpress (as well as the longpress timer).  Of course, if and when the flick distance threshold has been passed, we then lock the longpress out entirely - there will be a gesture stage "followup" mode for that in #9320.

Of course, the flick/longpress example is one we aim to sidestep for 17.0 by not trying to do them both ("resettable longpress" via roaming touch & "flick") at once.  That said, it is the clearest and most-direct example of the problem.   I recognized this during my first attempt, hence why the original version of this PR aimed to simply sidestep the issue entirely.  Alas, while the prior workaround - a single 'alias state' for each 'true source' - was fine at some levels, it has led to complications in upcoming ones that are more severe than before.

File: common/web/gesture-recognizer/src/engine/headless/simpleGestureSource.ts [hunk 2]
```diff
@@ -96,6 +100,18 @@ export class SimpleGestureSource<HoveredItemType> {
...
 
+  public constructSubview(startAtEnd: boolean, preserveBaseItem: boolean) {
+    return new SimpleGestureSourceSubview(this, startAtEnd, preserveBaseItem);
+  }
+
+  public terminate(cancel?: boolean) {
+    this.path.terminate(cancel);
+  }
+
+  public get isPathComplete(): boolean {
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/engine/headless/simpleGestureSource.ts:111]: Should we have doc comments on these? Especially given things that might not be obvious, like what does `cancel` mean for `terminate`?

**Review by mcdurdin — APPROVED** [review]
LGTM

**Pull request merged**